src/PlusDataCollection/OpenHaptics/vtkPlusOpenHapticsDevice.cxx

```diff
@@ -0,0 +1,297 @@
+  {
+    LOG_ERROR("vtkPlusOpenHapticsDevice::Probe should not be called while the device is already initialized");
+    return PLUS_FAIL;
+  }
+  if (this->InternalConnect())
+  {
+    this->InternalDisconnect();
+    LOG_INFO("vtkPlusOpenHapticsDevice::Probe found OpenHaptics Device");
+    return PLUS_SUCCESS;
+  }
+  else
+  {
+    LOG_ERROR("vtkPlusOpenHapticsDevice::Probe failed to connect to device");
+    return PLUS_FAIL;
+  }
+
+}
+
+PlusStatus vtkPlusOpenHapticsDevice::InternalUpdate()
+{
+  //schedule the internal callback - must be done each time device is polled
+  //the scheduler is intended to manage multiple clients requesting data from the same device
+  hdScheduleSynchronous(positionCallback, this, HD_MAX_SCHEDULER_PRIORITY);
+
+  return PLUS_SUCCESS;
+}
+
+PlusStatus vtkPlusOpenHapticsDevice::InternalConnect()
+{
+  if (this->isHapticDeviceInitialized)
+  {
+    LOG_DEBUG("Already connected - " << this->DeviceName);
+    return PLUS_SUCCESS;
+  }
+
+  if (this->InputChannels.empty())
+  {
+    LOG_WARNING("No force input has been provided");
+  }
+  else
+  {
+    if (this->InputChannels.size() > 1)
+    {
+      LOG_WARNING("Multiple input channels present.  Only first channel will be checked for force data");
+    }
+    vtkPlusDataSource* forceInput;
+    if (!this->InputChannels[0]->GetOwnerDevice()->GetToolByPortName("Force", forceInput))
+    {
+      LOG_WARNING("No Force tool in input channel.  Forces will not be available");
+    }
+  }
+
+  HDErrorInfo errorFlush;
+  while (HD_DEVICE_ERROR(errorFlush = hdGetError())) {}
+
+  DeviceHandle = hdInitDevice(this->DeviceName.c_str());
+
+  HDErrorInfo error;
+  if (HD_DEVICE_ERROR(error = hdGetError()))
+  {
+    LOG_ERROR( "Failed to initialize Phantom Omni " << this->DeviceName);
+    DeviceHandle = -1;
+    return PLUS_FAIL;
+  }
+
+  // Enable forces
+  hdEnable(HD_FORCE_OUTPUT);
+  hdEnable(HD_FORCE_RAMPING);
+
+  this->isHapticDeviceInitialized = true;
+  LOG_INFO("Phantom initialized: " << this->DeviceName)
+
+  hdStartScheduler();
+
+  return PLUS_SUCCESS;
+}
+
+PlusStatus vtkPlusOpenHapticsDevice::InternalDisconnect()
+{
+  if (this->isHapticDeviceInitialized)
+  {
+    hdDisableDevice(DeviceHandle);
+    hdStopScheduler();
+    this->isHapticDeviceInitialized=false;
+  }
+  return PLUS_SUCCESS;
+}
+
+PlusStatus vtkPlusOpenHapticsDevice::ReadConfiguration(vtkXMLDataElement* rootConfigElement)
+{
+  XML_FIND_DEVICE_ELEMENT_REQUIRED_FOR_READING(deviceConfig, rootConfigElement);
+  XML_READ_CSTRING_ATTRIBUTE_REQUIRED(DeviceName,deviceConfig);
+  return PLUS_SUCCESS;
+}
+
+PlusStatus vtkPlusOpenHapticsDevice::WriteConfiguration(vtkXMLDataElement* rootConfigElement)
+{
+  XML_FIND_DEVICE_ELEMENT_REQUIRED_FOR_WRITING(trackerConfig, rootConfigElement);
+  trackerConfig->SetAttribute("DeviceName", this->DeviceName.c_str());
+  return PLUS_SUCCESS;
+}
+
+PlusStatus vtkPlusOpenHapticsDevice::InternalStartRecording()
+{
+  if (!this->isHapticDeviceInitialized)

```

Comment:
If these functions do nothing, omit them as the parent level does basic checks for you.

Reply:
Thanks, will change this (and the related issues above).  This happened because I used the MicronTracker code as a starting template, and it is apparently out of date?

Reply:
Yes, the code has been evolving for a long time. Not everything gets cleaned up to the latest "normal".